tools/migration/internal/runner_test.go

```diff
@@ -1,26 +1,223 @@
 package internal_test
 
 import (
+	"bytes"
+	"errors"
 	"io/ioutil"
+	"log"
 	"os"
+	"path/filepath"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
+	"github.com/filecoin-project/go-filecoin/repo"
 	tf "github.com/filecoin-project/go-filecoin/testhelpers/testflags"
 	. "github.com/filecoin-project/go-filecoin/tools/migration/internal"
 )
 
-// TODO: Issue #2595 Implement first repo migration
 func TestMigrationRunner_Run(t *testing.T) {
 	tf.UnitTest(t)
-	dummyLogFile, err := ioutil.TempFile("", "logfile")
-	require.NoError(t, err)
-	logger := NewLogger(dummyLogFile, false)
-	defer func() {
-		require.NoError(t, os.Remove(dummyLogFile.Name()))
-	}()
-	runner := NewMigrationRunner(logger, "describe", "/home/filecoin-symlink")
-	assert.NoError(t, runner.Run())
+
+	repoDir, repoSymlink := RequireSetupTestRepo(t, 0)
+	defer RequireRemove(t, repoDir)
+	defer RequireRemove(t, repoSymlink)
+
+	t.Run("valid command returns error if repo not found", func(t *testing.T) {
+		dummyLogFile, dummyLogPath := RequireOpenTempFile(t, "logfile")
+		defer RequireRemove(t, dummyLogPath)
+		logger := NewLogger(dummyLogFile, false)
+		runner := NewMigrationRunner(logger, "describe", "/home/filecoin-symlink", "doesnt/matter")
+		assert.Error(t, runner.Run(), "no filecoin repo found in /home/filecoin-symlink.")
+	})
+
+	t.Run("can set MigrationsProvider", func(t *testing.T) {
+		dummyLogFile, dummyLogPath := RequireOpenTempFile(t, "logfile")
+		defer RequireRemove(t, dummyLogPath)
+		logger := NewLogger(dummyLogFile, false)
+		runner := NewMigrationRunner(logger, "describe", repoSymlink, "")
+		runner.MigrationsProvider = testProviderPasses
+
+		migrations := runner.MigrationsProvider()
+		assert.NotEmpty(t, migrations)
+		assert.NoError(t, runner.Run())
+	})
+
+	t.Run("Returns error and does not not run the migration if the repo is already up to date", func(t *testing.T) {
+		RequireSetRepoVersion(t, 1, repoDir)
+
+		dummyLogFile, dummyLogPath := RequireOpenTempFile(t, "logfile")
+		defer RequireRemove(t, dummyLogPath)
+		logger := NewLogger(dummyLogFile, false)
+		runner := NewMigrationRunner(logger, "describe", repoSymlink, "")
+		runner.MigrationsProvider = testProviderPasses
+		assert.EqualError(t, runner.Run(), "binary version 1 = repo version 1; migration not run")
+	})
+
+	t.Run("Runs the right migration version", func(t *testing.T) {
+		RequireSetRepoVersion(t, 1, repoDir)
+
+		dummyLogFile, dummyLogPath := RequireOpenTempFile(t, "logfile")
+		defer RequireRemove(t, dummyLogPath)
+		logger := NewLogger(dummyLogFile, false)
+		runner := NewMigrationRunner(logger, "describe", repoSymlink, "")
+		runner.MigrationsProvider = testProviderValidationFails
+		assert.EqualError(t, runner.Run(), "binary version 1 = repo version 1; migration not run")
```

Comment:
This looks like a copy-paste error from the previous test, judging from the t.run description.

Reply:
It was a leftover test idea I meant to delete.